Use webpack module federation to aggregate UIs and solutions UIs

JBWatenbergScality opened the pull request:
**Component**:

ui

**Context**: 

We are studying options to improve overall user experience while browsing the UI when several solutions UI are deployed and registered in the common navbar configuration.

**Summary**:

This PR introduces a set of changes to convert the common navbar to a bidirectional federated host (https://github.com/module-federation/module-federation-examples/tree/master/bi-directional) that can load UI remote bundle at runtime and hence be able to render an UI depending on a dynamic context where new solutions can be deployed / upgraded independently than the Metalk8s Shell-UI

**Acceptance criteria**: 

End2End tests, documentation and design docs should be released.

JBWatenbergScality committed:
shell-ui: Module federation wip

JBWatenbergScality committed:
ui: Module federation wip

JBWatenbergScality committed:
shell-ui: Refactor shell-ui to module federation. This allow runtime integration with others UIs and sharing of the authentication context as well as the theme

JBWatenbergScality committed:
shell-ui: Convert Alerts library to a set of hooks and component federable

JBWatenbergScality committed:
shell-ui: Temporarly add metalk8s-ui remote module config to navbar for demo purposes

JBWatenbergScality committed:
ui: Upgrade core-ui

JBWatenbergScality committed:
ui: Prepare rework to convert metalk8s-ui to a federable UI using webpack module federation

JBWatenbergScality committed:
ui/module-federation: Use federated useAlerts hook
Use ComponentWithFederatedImports which can load several federated hook
at the same time

JBWatenbergScality committed:
shell-ui: Bidirectional module federation host

JBWatenbergScality committed:
ui: Use federated useAuth hook to retrieve connected user infos

JBWatenbergScality committed:
ui: Use useTheme instead of theme defined in redux

bert-e commented:
# Hello jbwatenbergscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/shell-ui-switch-to-module-federation` and the
destination branch `development/2.10`.

Please resolve the conflict on **the feature branch** (`improvement/shell-ui-switch-to-module-federation`).

```bash
 $ git fetch
 $ git checkout origin/improvement/shell-ui-switch-to-module-federation
 $ git merge origin/development/2.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/shell-ui-switch-to-module-federation
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @nicolast

* @thomasdanan

* @lucieleonard

JBWatenbergScality committed:
ui: Use unique History for router accross ui mounts

JBWatenbergScality committed:
ui: Fix race condition when initializing salt

JBWatenbergScality committed:
ui: Fix duplicate fetches of nodes information

JBWatenbergScality committed:
ui: Force styled component to use the same version to be able to retireve the theme using useTheme hook accross the UIs

JBWatenbergScality committed:
ui: Extract alerts management from redux

JBWatenbergScality committed:
ui: Fix hook injection in components to avoid infinite rerendering

JBWatenbergScality committed:
shell : Fix alert context

JBWatenbergScality committed:
shell: Add an index file to shell-ui to debug remote entry

JBWatenbergScality committed:
ui: FederatedComponent useMemo to avoid useless re-renders and donot set state when component is unmounted in useDynamicScript hook

JBWatenbergScality committed:
shell: Implement deployed UI app provider

JBWatenbergScality committed:
ui: Init web fingers

JBWatenbergScality committed:
shell: Review shell config for module federation

JBWatenbergScality committed:
Shell: Add react-router to manage routing of federated pages in ShellUI

JBWatenbergScality committed:
shell: Add a static discovered list of deployed UIs

JBWatenbergScality committed:
shell: Use react-router to route to the federated app

JBWatenbergScality committed:
shell: Implement web fingerts providers

JBWatenbergScality committed:
module-federation: Add a context to detect which apps is the currently rendered one. This enables auto configuration of federated hooks

JBWatenbergScality committed:
shell: Federate an abstraction of useAuth to prepare support of OAuth2Poxy authentication

JBWatenbergScality committed:
shell: Federate an abstraction of useAuth to prepare support of OAuth2Poxy authentication

JBWatenbergScality committed:
shell: Shell configuration provider

JBWatenbergScality committed:
shell: Disable refetch of discovered UI and webfingers on focus

JBWatenbergScality committed:
ui: Add useVolumesWithAlerts hook to mutualize alert logic between volume page and volume tab in the node page

JBWatenbergScality committed:
ui: Remove unused navbar component

JBWatenbergScality committed:
shell: Add theme typing

JBWatenbergScality committed:
shell: Add support for federated and non federated navbar links

JBWatenbergScality committed:
shell: Remove deprecated code following to module federation switch

ChengYanJin committed:
ui/module-federation: Use intl.formatMessage to retrieve translations

Refs: ARTESCA 1556 + 1547

ChengYanJin committed:
ui/module-federation: Upgrade react-intl library

Refs: #2836

ChengYanJin committed:
ui/module-federation: Dispatch action to keep intl to redux

ChengYanJin committed:
ui/module-federation: Retrieve language from federated hook

- remove the intlGlobalProvider, retrieve intl from useIntl()
- extract intlProvider to load the federated hook useLanguage() thanks to
  `ComponentWithLazyHook`

ChengYanJin committed:
ui/theme: Remove the brand level in theme

Set the tooltip to tooltipTheme={'custom'} with the patch version of
core-ui v0.18.2 release

JBWatenbergScality committed:
ui: Create module federation library

JBWatenbergScality committed:
ui-module-federation: Replace ComponentWithLazyHooks by ComponentWithFederatedImports to make it support several imports and make it clear it can import other functions/modules than hooks

JBWatenbergScality committed:
ui,shell-ui: Use @scality/module-federation library

JBWatenbergScality committed:
shell: Display a loader while loading UI web fingers

JBWatenbergScality committed:
shell: Review tests folowing to module federation refactoring

ChengYanJin committed:
ui/tests: Add intlprovider for component unit test

JBWatenbergScality committed:
shell: Statically serve previous solution ui bundle to ensure retrocompatibility for solution UIs

ChengYanJin committed:
ui/tests: Skip the saga tests related to translations

JBWatenbergScality committed:
ui: Fix test on getVOlumeListData

JBWatenbergScality committed:
shell-ui: Clean log

ChengYanJin committed:
ui/theme: Fix the theme for ProgressBar

JBWatenbergScality committed:
ci: Module federation

ChengYanJin committed:
ui: Update with the patch v0.18.3 core-ui

JBWatenbergScality committed:
ui: Configure shell-ui remoteEntry url

ChengYanJin committed:
ui: Fix the dependency warning from useEffect hook

JBWatenbergScality committed:
ui,ci: Move UI module federation library to an independant repository

JBWatenbergScality committed:
ci: Adapt shell-ui ci to module federation refactoring

JBWatenbergScality committed:
test: Add mocks for UI service discovery in cypress tests

JBWatenbergScality committed:
test: Add mocks for runtime app configuration

JBWatenbergScality committed:
shell: Expose history to cypress

JBWatenbergScality committed:
salt: Increment shell-ui version and add merge scropt with older version

JBWatenbergScality committed:
shell: Fix shell config typing

JBWatenbergScality committed:
shell: Fix index.html target

JBWatenbergScality committed:
salt: Shell-UI configuration v1alpha2 mapped to runtime app configuration

JBWatenbergScality committed:
ui: Use config from runtime app config web finger instead of refetching it from /config.json

JBWatenbergScality committed:
salt: Runtime app configuration

JBWatenbergScality committed:
ui: Create voume, remove unnecessary intl.formatMessage

JBWatenbergScality committed:
salt: CSC framework for deployed ui apps

JBWatenbergScality committed:
shell: Webpack add entry.js in bundled index.html

JBWatenbergScality committed:
shell-ui: Bump module federation library

JBWatenbergScality committed:
ci: fixes

JBWatenbergScality committed:
ui: Use shell index.html for local dev

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @lucieleonard

* @nicolast

* @jbwatenbergscality

* @thomasdanan

* @gdemonet

* @alexandre-allard-scality

* @chengyanjin

* @teddyandrieux

JBWatenbergScality committed:
ui: Upgrade react-query to support module federation context sharing

JBWatenbergScality committed:
shell: Add an error boundary in case of error while rendering a micro app

JBWatenbergScality committed:
shell: Attach shared context to the window to enable micro apps to resue them

JBWatenbergScality committed:
shell: Import css and font

ChengYanJin committed:
shell-ui: Expose the UIListProvider module

ChengYanJin committed:
ui: Expose the platform library

JBWatenbergScality committed:
ui: Fix integration test

JBWatenbergScality committed:
ui: Fetch the app runtime config before fetching the basename

JBWatenbergScality committed:
shell,ui: bump module federation library

JBWatenbergScality committed:
docs: Update developer docs to run the UI locally

JBWatenbergScality committed:
changelog: Add an entry referring module federation changes

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbwatenbergscality

* @teddyandrieux

* @alexandre-allard-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @chengyanjin

* @nicolast

JBWatenbergScality committed:
ui: Expose /config.json configuration for backward compatibility

JBWatenbergScality committed:
shell: Introduce a ShellHistoryProvider to enable use of useLinkOpener in micor apps using react router

JBWatenbergScality committed:
ui,salt,shell: Add backward compatibility for alerts and platform libraries

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan

JBWatenbergScality committed:
shell: Fix selected tab for non federated entries

JBWatenbergScality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.10`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.10`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye jbwatenbergscality.
